File: Programmers_cpp/라면공장.cpp
#include <bits/stdc++.h>
using namespace std;

int solution(int stock, vector<int> dates, vector<int> supplies, int k) {
	priority_queue<int> pque;
	int answer = 0, index = 0;
	while (stock < k) {
		for (int i = index; i < dates.size(); i++) {
			if (dates[i] <= stock) {
				pque.push(supplies[i]);
			}
			else {
				index = i;
				break;
			}
		}
		stock += pque.top();
		pque.pop();
		answer++;
	}
	return answer;
}

// 출력 테스트를 위한 코드
int main(void) {
	ios::sync_with_stdio(false);
	cin.tie(0);

	int answer = solution(7, { 1,2,3,4 }, { 10,20,30,40 }, 30);
	cout << answer;
}

File: Programmers_java/라면공장.java
import java.util.Comparator;
import java.util.PriorityQueue;

public class 라면공장 {
	
    public static int solution(int stock, int[] dates, int[] supplies, int k) {
        int answer = 0, i = 0;
        PriorityQueue<Integer> que = new PriorityQueue<>(Comparator.reverseOrder());
        while (stock < k) {
			for (; i < dates.length; i++) {
				if(dates[i] <= stock) {
					que.add(supplies[i]);
				} else {
					break;
				}
			}
			stock += que.poll();
			answer++;
		}
        return answer;
    }
  
    
	public static void main(String[] args) {
		int[] dates = {1, 2, 3, 4};
		int[] supplies = {10, 40, 30 ,20};
		int ans = solution(4, dates, supplies, 100);
		System.out.println(ans);
	}
}
